proto/structured.proto

```diff
@@ -104,3 +105,234 @@ message CreateTableResponse {
+  // servicing the request will be set here. Additionally, in the case
+  // of writes, this value may be increased from the timestamp passed
+  // with the RequestHeader if the key being written was either read
+  // or written more recently.
+  //
+  // TODO(vivek): Is this only relevant in the non-transactional case?
+  optional Timestamp timestamp = 2 [(gogoproto.nullable) = false];
+  // Transaction is non-nil if the request specified a non-nil
+  // transaction. The transaction timestamp and/or priority may have
+  // been updated, depending on the outcome of the request.
+  optional Transaction txn = 3;
+}
+
+// A ColumnCollection represents a collection of columns as name:value pairs.
+message ColumnCollection {
+  message ColumnNameValue {
+    // Column name.
+    optional string name = 1;
+    optional MVCCValue value = 2 [(gogoproto.nullable) = false];
+  }
+  repeated ColumnNameValue columns = 1;
+}
+
+// A ValueCollection is a collection of MVCC values.
+message ValueCollection {
+  repeated MVCCValue values = 1;
+}
+
+// A GetTableRowRequest is used to read a table row.
+message GetTableRowRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional ColumnCollection key = 2 [(gogoproto.nullable) = false];
+  // Get specified columns from the row.
+  repeated string columns = 3;
+}
+
+// A GetTableRowResponse returns a row.
+// If the key doesn't exist, it returns an empty row.
+message GetTableRowResponse {
+  optional TableResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  // A deleted column has deleted set.
+  optional ValueCollection row = 2 [(gogoproto.nullable) = false];
+}
+
+// A PutTableRowRequest inserts/updates a row in the table. To write
+// an empty value in a cell specify an empty Bytes in the Value. To Delete
+// a column specify the deleted bit.
+message PutTableRowRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  // The columns forming the index key must be part of the ColumnCollection.
+  optional ColumnCollection row = 3 [(gogoproto.nullable) = false];
+}
+
+// A PutTableRowResponse is the response.
+message PutTableRowResponse {
+  optional TableResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
+
+// A ConditionalPutTableRowRequest.
+//
+// Puts row if exp_row equals an existing row in the DB. Both exp_row and row
+// must contain the columns forming the index key, and those constructed keys
+// must be equal. Returns error and the actual value of the row in the response
+// if exp_row doesn't match the full row in the DB.
+message ConditionalPutTableRowRequest {
+  optional TableRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional ColumnCollection exp_row = 3 [(gogoproto.nullable) = false];
+  optional ColumnCollection row = 4[(gogoproto.nullable) = false];
```

Comment:
Conditional put is an optimization. You can always implement it as a transaction. Should we provide it in the client protocol (or expose it in such a way) that the only implementation requires a transaction internally? Perhaps support for "atomic" conditional put of a row outside of a transaction is a good reason to avoid splitting a row across 2 ranges.